[ao][sparsity] Post training data sparsifier callback for lightning

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81813
* #80371
* __->__ #80370

Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

Commit:
[ao][sparsity] Post training data sparsifier callback for lightning

Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py```

[ghstack-poisoned]

HDCharles commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,62 @@
+
+    def on_validation_start(self, trainer, pl_module) -> None:
+        copied_model = deepcopy(pl_module.model).eval()  # do we really need to put in cpu?
+        self.sparsified = copied_model
```
why not assign directly

macandro96 replied:
done

HDCharles commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,62 @@
+            3. squashes the mask()
+    """
+
+    def on_validation_start(self, trainer, pl_module) -> None:
```
don't think this is the correct place for this, shouldn't it be on_fit_end for PTS?

macandro96 replied:
changed this from to on_fit_end()

HDCharles commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:
```diff
@@ -0,0 +1,32 @@
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+def _create_data_sparsifier(data_sparsifier_args, data_sparsifier_type):
```
this function feels a little bit redundant if it just takes 2 arguments and calls one with the other as an arg.

if not, should probably have arg order swapped to be more intuitive

macandro96 replied:
yeah, seems redundant - directly creating the object now

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

macandro96 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -0,0 +1,84 @@
+
+
+
+if __name__ == "__main__":
```
modified the test structure

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,59 @@
+import torch
+from dataclasses import dataclass, field
+from typing import Any, Optional, Dict
+import pytorch_lightning as pl
```
```
import pytorch_lightning as pl  # type: ignore
```

macandro96 replied:
This still seems to throw TYPEIGNORE error. I get this 
```This line has an unqualified `type: ignore`; please convert it to `type:
    ignore[xxxx]```

macandro96 replied:
```type: ignore[import]``` seemed to fix the issue.

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,59 @@
+        <callback_obj>.sparsified
+
+    Args:
+        data_sparsifier_type (some implemented class of BaseDataSparsifier)
```
maybe just `data_sparsifier_class`? I feel type is a bit ambiguous

macandro96 replied:
Yeah, makes sense. Changed it to data_sparsifier_class for now.

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,59 @@
+            Note: Objects should not be passed in here as they are created
+            once the training completes.
+
+        data_sparsifier_args (Dict)
```
wondering if you would like to adopt the `with_args` design: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/observer.py#L1460

jerryzh168 replied:
in this case `data_sparsifier_type` should be named as `data_sparsifier_constructor` I think

jerryzh168 commented on torch/ao/sparsity/experimental/data_sparsifier/lightning/callbacks/data_sparsity.py:
```diff
@@ -0,0 +1,59 @@
+
+    def on_fit_end(self, trainer, pl_module) -> None:
+        self.sparsified = deepcopy(pl_module.model).eval()
+        self.data_sparsifier = self.data_sparsifier_type(**self.data_sparsifier_args)
```
also I'm wondering if we can just pass in a data_sparsifier directly from user?

macandro96 replied:
We could do that for the data_sparsifier, but I don't think we can for the scheduler. Because the data_scheduler is created after the sparsifier is attached to the model.
So, just to keep things simple, we ask the user to input only the class_name and not the object for both data_sparsifier and data_scheduler.

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -0,0 +1,89 @@
+            i = c
+
+
+class DummyLightningModule(pl.LightningModule):
+    def __init__(self, iC: int, oC: List[int]):
+        super().__init__()
+        self.model = DummyModel(iC, oC)
+
+    def forward(self):
+        pass
```
Let's wrap this into a creator function to encapsulate the pl import? Something like:

```
def make_lightning_module(*args, **kwargs):
    import pytorch_lightning as pl # type: ignore[import]
    class DummyLightningModule(pl.LightningModule):
        ...
    return DummyLightningModule(*args, **kwargs)
```

macandro96 replied:
yep

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -0,0 +1,89 @@
+            'zeros_per_block': 4
+        }
+        callback = PostTrainingDataSparsity(DataNormSparsifier, sparsifier_args)
+        pl_module = DummyLightningModule(100, [128, 256, 16])
```
If using a creator function, you can replace this with

```
pl_module = make_lightning_module(100, [128, 256, 16])
```

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -0,0 +1,89 @@
+
+
+class TestPostTrainingCallback(_PostTrainingCallbackTestCase):
+    def test_post_training_callback(self):
```
Can we add a decorator to skip? Something like

```
@unittest.skipIf(not importlib.util.find_spec("pytorch_lightning"), "No pytorch_lightning")
def test_post_training_callback(self):
    ...
```

Double check if the skipping works :)) I didn't confirm lolol

macandro96 replied:
I did - seems like all the tests pass now

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/callbacks/_data_sparstity_utils.py:
```diff
@@ -0,0 +1,36 @@
+    Essentialy, loop over all the weight parameters in the module and
+    attach it to the data sparsifier.
+    Note::
+        The '.' in the layer names are replaced with some other character
+        before attaching to the sparsifier. This is because, the data
+        sparsifier uses a dummy model inside to store the weight parameters.
```
Be more specific: what do you mean "with some other character"?

z-a-f replied:
Also, I don't think a layer name can have a `.` in there. Do you support FQNs?

macandro96 replied:
Yes, the layer name cannot have a ```.```. We're basically taking the name of the model parameter and replacing it with ```_```.

z-a-f commented on torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py:
```diff
@@ -0,0 +1,94 @@
+    return DummyLightningModule(iC, oC)
+
+
+class _PostTrainingCallbackTestCase(TestCase):
```
I still don't see the point of creating a separate class for the test case in this scenario

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g

Commit:
Update on "[ao][sparsity] Post training data sparsifier callback for lightning"


Lightning callback that enables post-training sparsity.

This callback aims to sparsify the model inside lightning module after training.
**Note that the model is copied and then sparsified, so the existing model is not modified**

The sparsified model can be used for comparison and can be accessed using <callback_obj>.sparsified

Test Plan
```python torch/ao/sparsity/_experimental/data_sparsifier/lightning/tests/test_callbacks.py TestPostTrainingCallback```

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
